### PR TITLE
Move impl constness into impl trait header

### PR DESCRIPTION
**Pull request by oli-obk**
This PR is kind of the opposite of the rejected https://github.com/rust-lang/rust/pull/134114

Instead of moving more things into the `constness` query, we want to keep them where their corresponding hir nodes are lowered. So I gave this a spin for impls, which have an obvious place to be (the impl trait header). And surprisingly it's also a perf improvement (likely just slightly better query & cache usage).

The issue was that removing anything from the `constness` query makes it just return `NotConst`, which is wrong. So I had to change it to `bug!` out if used wrongly, and only then remove the impl blocks from the `constness` query. I think this change is good in general, because it makes using `constness` more robust (as can be seen by how few sites that had to be changed, so it was almost solely used specifically for the purpose of asking for functions' constness). The main thing where this change was not great was in clippy, which was using the `constness` query as a general DefId -> constness map. I added a `DefKind` filter in front of that. If it becomes a more common pattern we can always move that helper into rustc.

### CODE REVIEW

**Comment by oli-obk** [comment]
yay

**Comment by petrochenkov** [comment]
The implementation looks good to me, but from the description it's still not clear to me what is the motivation for this part

>then remove the impl blocks from the `constness` query

(moving impl constness from one query to another).

Also one of the tests still hits the assert.

@rustbot author

**Comment by oli-obk** [comment]
@rustbot ready

I added a motivation section.

The clippy test failure was actually a counterargument for this change, but since it's only a single clippy lint, I think the change in itself is still net positive.

**Comment by petrochenkov** [comment]
cc @fee1-dead as the reviewer of #134114.

@bors r+

File: compiler/rustc_const_eval/src/const_eval/fn_queries.rs [hunk 1]
```diff
@@ -4,10 +4,15 @@ use rustc_hir::def_id::{DefId, LocalDefId};
...
+    if matches!(tcx.def_kind(parent_id), DefKind::Impl { .. })
+        && let Some(header) = tcx.impl_trait_header(parent_id)
+    {
+        header.constness
+    } else {
+        hir::Constness::NotConst
+    }
 }
 
 /// Checks whether an item is considered to be `const`. If it is a constructor, anonymous const,
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/fn_queries.rs:18]: This comment looks like is still needs updating.

**Comment by Kobzol** [comment]
More improvements than regressions.

@rustbot label: +perf-regression-triaged

**Pull request merged**